File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        List<Integer> points = new ArrayList<>();
        int before = -1, idx = 0;

        while (head != null && head.next != null) {
            if (before != -1 && (
                (head.val < before && head.val < head.next.val) ||
                (head.val > before && head.val > head.next.val)
            )) points.add(idx);

            idx++;
            before = head.val;
            head = head.next;
        }

        if (points.size() < 2) return new int[]{-1, -1};
        int[] result = {Integer.MAX_VALUE, 0};

        for (int i = 1; i < points.size(); i++) {
            result[0] = Math.min(result[0], points.get(i) - points.get(i - 1));
            result[1] = Math.max(result[1], points.get(i) - points.get(0));
        }

        return result;
    }
}

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        points = list()
        before = -1
        idx = 0

        while head and head.next:
            if (before != -1 and (
                (head.val < before and head.val < head.next.val) or
                (head.val > before and head.val > head.next.val))):
                points.append(idx)

            idx += 1
            before = head.val
            head = head.next

        if len(points) < 2: return [-1, -1]
        result = [int(1e12), 0]

        for i in range(1, len(points)):
            result[0] = min(result[0], points[i] - points[i - 1])
            result[1] = max(result[1], points[i] - points[0])

        return result